File: app/src/main/java/com/billpayments/features/payments/data/PaymentRepository.kt
package com.billpayments.features.payments.data

import com.billpayments.core.mock.MockNetwork

interface PaymentRepository {
    suspend fun pay(payment: Payment)
}

class FakePaymentRepository(private val network: MockNetwork) : PaymentRepository {

    override suspend fun pay(payment: Payment) {
        network.delay()
        network.countAndMaybeFail()
    }
}

File: BillPayments/Features/Payments/Data/PaymentRepository.swift
protocol PaymentRepository {
    func pay(_ payment: Payment) async throws
}

final class FakePaymentRepository: PaymentRepository {
    private let network: MockNetwork

    init(network: MockNetwork) {
        self.network = network
    }

    func pay(_ payment: Payment) async throws {
        try await network.delay()
        try network.countAndMaybeFail()
    }
}
